crates/bevy_ecs/src/spawn.rs

```diff
@@ -282,16 +283,26 @@ unsafe impl<R: Relationship, L: SpawnableList<R> + Send + Sync + 'static> Bundle
     }
 }
 
-impl<R: Relationship, L: SpawnableList<R>> DynamicBundle for SpawnRelatedBundle<R, L> {
+unsafe impl<R: Relationship, L: SpawnableList<R>> DynamicBundle for SpawnRelatedBundle<R, L> {
     type Effect = Self;
 
-    fn get_components(
-        self,
+    unsafe fn get_components(
+        ptr: *const Self,
         func: &mut impl FnMut(crate::component::StorageType, bevy_ptr::OwningPtr<'_>),
-    ) -> Self::Effect {
-        <R::RelationshipTarget as RelationshipTarget>::with_capacity(self.list.size_hint())
-            .get_components(func);
-        self
+    ) {
+        // SAFETY: The caller must ensure that the `ptr` must be valid but not necessarily aligned.
+        let effect = ManuallyDrop::new(unsafe { ptr.read() });
+        let target =
+            <R::RelationshipTarget as RelationshipTarget>::with_capacity(effect.list.size_hint());
+        let target = ManuallyDrop::new(target);
+        let target_ptr = &raw const target;
+        <R::RelationshipTarget as DynamicBundle>::get_components(target_ptr.cast::<R::RelationshipTarget>(), func);
+    }
+
+    unsafe fn apply_effect(ptr: *const Self, entity: &mut EntityWorldMut) {
+        // SAFETY: The caller must ensure that the `ptr` must be valid but not necessarily aligned.
+        let effect = unsafe { ptr.read() };
+        effect.apply(entity);
```

Comment:
This is unsound.

`get_components` is moving `effect.list` to a local variable and then calls a method on it. Even though it takes `&self` it can still modify the local instance by using internal mutability. If this happens then `apply_effect` will read a stale copy of `effect.list`, which should not happen.

You have to either call `size_hint` directly on the instance pointed by `ptr` (assuming it is aligned) or write back `effect.list` to `ptr` after calling `size_hint` on it. Or just don't pick an initial capacity, avoiding the need to read `effect` in `get_components` in the first place.

Reply:
I've changed this to turn the pointer in `get_components` into a borrow, as it fulfills all of the requirements for one.